### DISCUSSION TITLE
Unresolved TypeVar: How to provide overload when kwarg not provided

### DISCUSSION THREAD

**Discussion by jonaslb** [microsoft/pyright#8998] [Q&A] [answered]
As an example, see the following stub (`.pyi`) file:

```
from typing import TypeVar, overload

T = TypeVar("T")

@overload
def func(*, x: int = ...) -> None: ...
@overload
def func(*, x: int = ..., t: T = ...) -> list[T]: ...
```

Pyright complains that `T` may go unresolved on the last line, and to provide an overload for when the argument is not provided (the `reportInvalidTypeVarUse` lint). However, that's precisely what I tried to do without success.

Is this a bug or how would I specify this correctly?

**Comment by erictraut** [comment, marked as answer]
You can delete the `= ...` in the second overload. The first overload already covers the case where a keyword argument `t` is not provided.

**Reply by jonaslb** [reply]
Ah, I don't know why I didn't realise that, that's clearly correct. However it touches on a mistake I made in constructing the example :( If I can make a follow-up question - say there is no `*`, i.e. these can be both positional or keyword arguments - I can't put `t: T` after `x: int = ...` in the argument list due to `Non-default argument follows default argument`.  I don't want to construct overrides for whether `x` is provided or not, since I actually have multiple other optional arguments before the one I want to add (generic) typing to, so I imagine there would be some combinatorial complexity in doing that.

**Reply by jonaslb** [reply]
On second thought it might not be as horrible as I had initially thought. If `t` is positional all the other args must be too (as it it the last one), so it's only one possibility. On the other hand, if it is given as a kwarg, it could be moved to the right side of `*` and still be marked obligatory. Thanks for pointing out what I couldn't see anyway - although I think it would be easier for me if it was possible to do what I tried in the question (minus the kwarg only stuff).
